crates/orchestrator/src/network_spec/relaychain.rs

```diff
@@ -0,0 +1,118 @@
+    pub(crate) default_resources: Option<Resources>,
+
+    /// Default database snapshot. Can be overriden on each node.
+    pub(crate) default_db_snapshot: Option<AssetLocation>,
+
+    /// Default arguments to use in nodes. Can be overriden on each node.
+    pub(crate) default_args: Vec<Arg>,
+
+    // chain_spec_path: Option<AssetLocation>,
+    pub(crate) chain_spec: ChainSpec,
+
+    /// Set the count of nominators to generator (used with PoS networks).
+    pub(crate) random_nominators_count: u32,
+
+    /// Set the max nominators value (used with PoS networks).
+    pub(crate) max_nominations: u8,
+
+    /// Nodes to run.
+    pub(crate) nodes: Vec<NodeSpec>,
+}
+
+impl RelaychainSpec {
+    pub fn from_config(config: &RelaychainConfig) -> Result<RelaychainSpec, OrchestratorError> {
+        // Relaychain main command to use, in order:
+        // set as `default_command` or
+        // use the command of the first node.
+        // If non of those is set, return an error.
+        let main_cmd = if let Some(cmd) = config.default_command() {
+            cmd
+        } else if let Some(first_node) = config.nodes().first() {
+            let Some(cmd) = first_node.command() else {
+                return Err(OrchestratorError::InvalidConfig("Relaychain, either default_command or command in the first node needs to be set.".to_string()));
+            };
+
+            cmd
+        } else {
+            return Err(OrchestratorError::InvalidConfig(
+                "Relaychain without nodes and default_command isn't set.".to_string(),
+            ));
+        };
+
+        let chain_spec = ChainSpec::new(config.chain().as_str(), Context::Relay)
+            .set_chain_name(config.chain().as_str());
+        let chain_spec = if let Some(chain_spec_path) = config.chain_spec_path() {
+            chain_spec.asset_location(chain_spec_path.clone())
+        } else {
+            // TODO: Do we need to add the posibility to set the command to use?
+            // Currently (v1) is possible but when is set is set to the default command.
+            chain_spec.commad(main_cmd.as_str())
+        };
+
+        // build the `node_specs`
+        let chain_context = ChainDefaultContext {
+            default_command: config.default_command(),
+            default_image: config.default_image(),
+            default_resources: config.default_resources(),
+            default_db_snapshot: config.default_db_snapshot(),
+            default_args: config.default_args(),
+        };
+
+        // We want to track the errors for all the nodes and report them ones
+        let mut errs: Vec<OrchestratorError> = Default::default();
+
+        let mut nodes: Vec<NodeSpec> = Default::default();
+        config.nodes().iter().for_each(|node_config| {
+            match NodeSpec::from_config(node_config, &chain_context) {
+                Ok(node) => nodes.push(node),
+                Err(err) => errs.push(err),
+            }
+        });
```

Comment:
nit: One really cool trick to make code more declarative is to use ```.fold``` in conjunction with ```.map```, that way we can create the NodeSpec and collect error at the same time

```suggestion
        let (nodes, errs) = config
            .nodes()
            .iter()
            .map(|node_config| NodeSpec::from_config(node_config, &chain_context))
            .fold((vec![], vec![]), |(mut nodes, mut errs), result| {
                match result {
                    Ok(node) => nodes.push(node),
                    Err(err) => errs.push(err),
                }
                (nodes, errs)
            });
```

Reply:
Yes, sounds good!